File: simulator/backend/services/parameter-service.ts
/**
 * Parameter Service
 * Business logic for parameter operations
 */

import { parameterRepository } from '../db/repositories/parameter-repository.ts';
import type { ParameterData, CreateParameterRequest, UpdateParameterRequest } from '@/shared/types/api.types';

export class ParameterService {
  /**
   * Get parameter by ID
   */
  async getParameterById(id: string): Promise<ParameterData | null> {
    try {
      return await parameterRepository.findById(id);
    } catch (error) {
      console.error('Error getting parameter by ID:', error);
      throw error;
    }
  }

  /**
   * Get all parameters
   */
  async getAllParameters(): Promise<ParameterData[]> {
    try {
      return await parameterRepository.findAll();
    } catch (error) {
      console.error('Error getting all parameters:', error);
      throw error;
    }
  }

  /**
   * Get parameters by type
   */
  async getParametersByType(type: 'structured' | 'narrative' | 'guardrails'): Promise<ParameterData[]> {
    try {
      return await parameterRepository.findByType(type);
    } catch (error) {
      console.error('Error getting parameters by type:', error);
      throw error;
    }
  }

  /**
   * Get parameters by category
   */
  async getParametersByCategory(categoryId: string): Promise<ParameterData[]> {
    try {
      return await parameterRepository.findByCategoryId(categoryId);
    } catch (error) {
      console.error('Error getting parameters by category:', error);
      throw error;
    }
  }

  /**
   * Create parameter
   */
  async createParameter(parameterData: CreateParameterRequest): Promise<ParameterData> {
    try {
      // Validate parameter data
      this.validateParameter(parameterData);

      return await parameterRepository.create(parameterData);
    } catch (error) {
      console.error('Error creating parameter:', error);
      throw error;
    }
  }

  /**
   * Update parameter
   */
  async updateParameter(id: string, parameterData: UpdateParameterRequest): Promise<ParameterData> {
    try {
      // Check if parameter exists
      const existing = await parameterRepository.findById(id);
      if (!existing) {
        throw new Error('Parameter not found');
      }

      return await parameterRepository.update(id, parameterData);
    } catch (error) {
      console.error('Error updating parameter:', error);
      throw error;
    }
  }

  /**
   * Delete parameter
   */
  async deleteParameter(id: string): Promise<void> {
    try {
      const existing = await parameterRepository.findById(id);
      if (!existing) {
        throw new Error('Parameter not found');
      }

      await parameterRepository.delete(id);
    } catch (error) {
      console.error('Error deleting parameter:', error);
      throw error;
    }
  }

  /**
   * Get all parameter categories
   */
  async getAllCategories(): Promise<any[]> {
    try {
      return await parameterRepository.getAllCategories();
    } catch (error) {
      console.error('Error getting parameter categories:', error);
      throw error;
    }
  }

  /**
   * Reset parameters to defaults
   */
  async resetToDefaults(): Promise<void> {
    try {
      await parameterRepository.resetToDefaults();
    } catch (error) {
      console.error('Error resetting parameters:', error);
      throw error;
    }
  }

  /**
   * Get parameters as key-value object
   */
  async getParametersAsObject(type?: 'structured' | 'narrative' | 'guardrails'): Promise<Record<string, any>> {
    try {
      const params = type
        ? await parameterRepository.findByType(type)
        : await parameterRepository.findAll();

      const paramObj: Record<string, any> = {};
      params.forEach(p => {
        try {
          paramObj[p.name] = typeof p.value === 'string' ? JSON.parse(p.value) : p.value;
        } catch {
          paramObj[p.name] = p.value;
        }
      });

      return paramObj;
    } catch (error) {
      console.error('Error getting parameters as object:', error);
      throw error;
    }
  }

  /**
   * Validate parameter data
   */
  private validateParameter(parameterData: CreateParameterRequest): void {
    if (!parameterData.name) {
      throw new Error('Parameter name is required');
    }

    if (!parameterData.type) {
      throw new Error('Parameter type is required');
    }

    if (!['structured', 'narrative', 'guardrails'].includes(parameterData.type)) {
      throw new Error('Invalid parameter type');
    }

    if (parameterData.value === undefined || parameterData.value === null) {
      throw new Error('Parameter value is required');
    }
  }

  /**
   * Export parameters to JSON
   */
  async exportParameters(): Promise<string> {
    try {
      const parameters = await parameterRepository.findAll();
      return JSON.stringify(parameters, null, 2);
    } catch (error) {
      console.error('Error exporting parameters:', error);
      throw error;
    }
  }

  /**
   * Import parameters from JSON
   */
  async importParameters(jsonData: string): Promise<void> {
    try {
      const parameters = JSON.parse(jsonData);

      if (!Array.isArray(parameters)) {
        throw new Error('Invalid parameter data format');
      }

      // Reset existing parameters
      await parameterRepository.resetToDefaults();

      // Import new parameters
      for (const param of parameters) {
        await parameterRepository.create(param);
      }
    } catch (error) {
      console.error('Error importing parameters:', error);
      throw error;
    }
  }
}

// Export singleton instance
export const parameterService = new ParameterService();

File: backend-python/app/services/parameter_service.py
"""
Parameter Service
Business logic for parameter catalog operations
"""

import logging
from typing import Any

from app.repositories.parameter_repository import parameter_repository
from app.models.parameter import ParameterData, CreateParameterRequest, UpdateParameterRequest

logger = logging.getLogger(__name__)


class ParameterService:
    """Service for parameter catalog operations."""

    async def get_parameter_by_id(self, parameter_id: str) -> ParameterData | None:
        """Get parameter by ID."""
        try:
            return await parameter_repository.find_by_id(parameter_id)
        except Exception as e:
            logger.error(f"Error getting parameter by ID: {e}")
            raise

    async def get_parameter_by_name(self, name: str) -> ParameterData | None:
        """Get parameter by name."""
        try:
            return await parameter_repository.find_by_name(name)
        except Exception as e:
            logger.error(f"Error getting parameter by name: {e}")
            raise

    async def get_all_parameters(self) -> list[ParameterData]:
        """Get all parameters."""
        try:
            return await parameter_repository.find_all()
        except Exception as e:
            logger.error(f"Error getting all parameters: {e}")
            raise

    async def get_parameters_by_type(self, param_type: str) -> list[ParameterData]:
        """Get parameters by type."""
        try:
            return await parameter_repository.find_by_type(param_type)
        except Exception as e:
            logger.error(f"Error getting parameters by type: {e}")
            raise

    async def get_parameters_by_category(self, category: str) -> list[ParameterData]:
        """Get parameters by category."""
        try:
            return await parameter_repository.find_by_category(category)
        except Exception as e:
            logger.error(f"Error getting parameters by category: {e}")
            raise

    async def get_active_parameters(self) -> list[ParameterData]:
        """Get active parameters."""
        try:
            return await parameter_repository.find_active()
        except Exception as e:
            logger.error(f"Error getting active parameters: {e}")
            raise

    async def create_parameter(self, param_data: CreateParameterRequest | dict[str, Any]) -> ParameterData:
        """Create new parameter."""
        try:
            return await parameter_repository.create(param_data)
        except Exception as e:
            logger.error(f"Error creating parameter: {e}")
            raise

    async def update_parameter(
        self,
        parameter_id: str,
        param_data: UpdateParameterRequest | dict[str, Any],
    ) -> ParameterData:
        """Update parameter."""
        try:
            return await parameter_repository.update(parameter_id, param_data)
        except Exception as e:
            logger.error(f"Error updating parameter: {e}")
            raise

    async def delete_parameter(self, parameter_id: str) -> None:
        """Delete parameter."""
        try:
            await parameter_repository.delete(parameter_id)
        except Exception as e:
            logger.error(f"Error deleting parameter: {e}")
            raise

    async def get_categories(self) -> list[str]:
        """Get all unique categories."""
        try:
            return await parameter_repository.get_categories()
        except Exception as e:
            logger.error(f"Error getting categories: {e}")
            raise

    async def get_parameter_catalog(self) -> dict[str, list[dict[str, Any]]]:
        """Get full parameter catalog grouped by type."""
        try:
            all_params = await parameter_repository.find_all()
            catalog = {
                "structured": [],
                "narrative": [],
                "guardrail": [],
            }

            for param in all_params:
                param_dict = param.model_dump()
                if param.type in catalog:
                    catalog[param.type].append(param_dict)

            return catalog
        except Exception as e:
            logger.error(f"Error getting parameter catalog: {e}")
            raise

    async def reset_parameters(self, param_type: str | None = None) -> None:
        """Reset parameters (delete all of a type or all)."""
        try:
            if param_type:
                await parameter_repository.delete_by_type(param_type)
            else:
                # Delete all parameters
                for ptype in ["structured", "narrative", "guardrail"]:
                    await parameter_repository.delete_by_type(ptype)
        except Exception as e:
            logger.error(f"Error resetting parameters: {e}")
            raise


# Singleton instance
parameter_service = ParameterService()
